core/pom.xml

```diff
@@ -40,6 +40,7 @@ THE SOFTWARE.
   <properties>
     <staplerFork>true</staplerFork>
     <hamcrest.version>2.2</hamcrest.version>
+    <junit.jupiter.version>5.6.2</junit.jupiter.version>
```

Comment:
I take it you have not tried to convert the `test` module with `JenkinsRule`? That is where most test coverage lies, but I wonder if a migration is practical.

TBH other than `assertThrows` (which could equally be a static utility method anywhere and used with JUnit 4) I have not seen a compelling advantage to using JUnit 5. It does not correct the main sin of JUnit 4, which is the abuse of annotations.

Reply:
`assertThrows` is in junit 4.13

Reply:
Actually I took a look at the ``test`` module, but this might take a little longer, because the JenkinsRule has to be changed to an Extension. If this would be needed before this PR can be merged, I can spend there some time.

But to answer the question about JUnit5 in general: I think there are a few things which are easier with JUnit5 than JUnit4, but that is only my opinion. Some examples:

- [Tests enabled for a specified JRE range ](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/condition/EnabledForJreRange.html)
- dynamic tests with @TestFactory

Another nice blog I stumbled upon is this one: [why you should migrate to junit-5](https://medium.com/riag-digital-techblog/why-you-should-migrate-to-junit-5-15438f5c56f)

Reply:
> If this would be needed before this PR can be merged

No, just wondering if that is part of the plan.

`@TestFactory` does sound easier to work with than JUnit 4 parameterized tests.

My experience with tests in @jenkinsci is that it is tough to make various systems interoperate cleanly, due to the misguided attempt by test frameworks to make things look declarative using annotations and reflection. Thus there are problems mixing `JenkinsRule` or `RestartableJenkinsRule` with timeouts, abstract base classes, `LoggerRule`, `docker-fixtures`, before/after class/method, parameters, ad nauseam. As a developer I know exactly what I want to happen in what order but the framework, or perhaps misuse of the framework, is an obstacle rather than a convenience.

Reply:
@jglick would you approve it after the discussion?

Reply:
I was never blocking anything here, just discussing possibilities and plans.